test/junit/scala/collection/mutable/OpenHashMapTest.scala

```diff
@@ -39,4 +49,50 @@ class OpenHashMapTest {
     // TODO assertEquals(0, fieldMirror.get.asInstanceOf[Int])
     assertEquals(0, field.getInt(m))
   }
+
+  /** Test that an [[OpenHashMap]] frees references to a deleted key (SI-9522). */
+  @Test
+  def freesDeletedKey {
+    import scala.language.reflectiveCalls
+
+    class MyClass {
+      override def hashCode() = 42
+    }
+
+    val counter = new GraphVisitor() {
+      private[this] var instanceCount: Int = _
+
+      def countInstances(obj: AnyRef) = {
+        instanceCount = 0
+        val walker = new GraphWalker(obj)
+        walker.addVisitor(this)
+        walker.walk
+        instanceCount
+      }
+
+      override def visit(record: GraphPathRecord) {
+        if (record.klass() == classOf[MyClass])  instanceCount += 1
+      }
+    }
```

Comment:
I was going to suggest to put a utility for this in the `scala.tools.testing` package (in test/junit), then I found the existing method [`assertNotReachable` in `scala.tools.testing.AssertUtil`](https://github.com/scala/scala/blob/95f57600a7498f269d32743788c89d4d65f93700/test/junit/scala/tools/testing/AssertUtil.scala#L68). Its internal method `refs` collects all objects reachable from a certain root, so you could factor this out and use it (or a new utility) in this test. It would be really nice if we could avoid the new dependency on org.openjdk.jol.